src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+        visitAnnotations(f.invisibleAnnotations)
+        visitAnnotations(f.invisibleTypeAnnotations)
+        visitFieldSignature(f.signature)
+      }
+
+      for (m <- classNode.methods.asScala) {
+        visitDescriptor(m.desc)
+
+        visitAnnotations(m.visibleAnnotations)
+        visitAnnotations(m.visibleTypeAnnotations)
+        visitAnnotations(m.invisibleAnnotations)
+        visitAnnotations(m.invisibleTypeAnnotations)
+        visitAnnotationss(m.visibleParameterAnnotations)
+        visitAnnotationss(m.invisibleParameterAnnotations)
+        visitAnnotations(m.visibleLocalVariableAnnotations)
+        visitAnnotations(m.invisibleLocalVariableAnnotations)
+
+        m.exceptions.asScala foreach visitInternalName
+        for (tcb <- m.tryCatchBlocks.asScala) visitInternalName(tcb.`type`)
+
+        val iter = m.instructions.iterator()
+        while (iter.hasNext) iter.next() match {
+          case ti: TypeInsnNode           => visitInternalNameOrArrayReference(ti.desc)
+          case fi: FieldInsnNode          => visitInternalNameOrArrayReference(fi.owner); visitDescriptor(fi.desc)
+          case mi: MethodInsnNode         => visitInternalNameOrArrayReference(mi.owner); visitDescriptor(mi.desc)
+          case id: InvokeDynamicInsnNode  => visitDescriptor(id.desc); visitHandle(id.bsm); id.bsmArgs foreach visitConstant
+          case ci: LdcInsnNode            => visitConstant(ci.cst)
+          case ma: MultiANewArrayInsnNode => visitDescriptor(ma.desc)
+          case _ =>
+        }
+
+        visitMethodSignature(m.signature)
+      }
+    }
+
+    def visitInternalName(internalName: InternalName): Unit = if (internalName != null) {
+      for (c <- getClassIfNested(internalName))
+        innerClasses += c
+    }
+
+    // either an internal/Name or [[Linternal/Name; -- there are certain references in classfiles
+    // that are either an internal name (without the surrounding `L;`) or an array descriptor
+    // `[Linternal/Name;`.
+    def visitInternalNameOrArrayReference(ref: String): Unit = if (ref != null) {
+      val bracket = ref.lastIndexOf('[')
+      if (bracket == -1) visitInternalName(ref)
+      else if (ref.charAt(bracket + 1) == 'L') visitInternalName(ref.substring(bracket + 2, ref.length - 1))
+    }
+
+    // we are only interested in the class references in the descriptor, so we can skip over
+    // primitives and the brackets of array descriptors
+    def visitDescriptor(desc: String): Unit = (desc.charAt(0): @switch) match {
+      case '(' =>
+        val internalNames = mutable.ListBuffer.empty[String]
+        var i = 1
+        while (i < desc.length) {
+          if (desc.charAt(i) == 'L') {
+            val start = i + 1 // skip the L
+            while (desc.charAt(i) != ';') i += 1
+            internalNames += desc.substring(start, i)
+          }
+          // skips over '[', ')', primitives
+          i += 1
+        }
+        internalNames foreach visitInternalName
```

Comment:
Can we avoid building up the collection here and just directly call `visitInternalName` immediately after we find each class name?